contracts/MuonNodeStaking.sol

```diff
@@ -0,0 +1,624 @@
+            "The sender is not the owner of the NFT."
+        );
+
+        bondedToken.transferFrom(msg.sender, address(this), tokenIdA);
+        bondedToken.approve(address(bondedToken), tokenIdA);
+
+        bondedToken.merge(tokenIdA, tokenIdB);
+
+        updateStaking();
+    }
+
+    /**
+     * @dev Calculates the total value of a bonded token in terms of the staking tokens.
+     * @param tokenId The id of the bonded token.
+     * @return amount The total value of the bonded token.
+     */
+    function valueOfBondedToken(uint256 tokenId)
+        public
+        view
+        returns (uint256 amount)
+    {
+        uint256[] memory lockedAmounts = bondedToken.getLockedOf(
+            tokenId,
+            stakingTokens
+        );
+
+        amount = 0;
+        for (uint256 i = 0; i < lockedAmounts.length; i++) {
+            address token = stakingTokens[i];
+            uint256 multiplier = stakingTokensMultiplier[token];
+            amount += (multiplier * lockedAmounts[i]) / 1e18;
+        }
+        return amount;
+    }
+
+    /**
+     * @dev Updates the staking status for the staker.
+     * This function calculates the staked amount based on the locked tokens and their multipliers,
+     * and updates the balance and total staked amount accordingly.
+     * Only callable by staker.
+     */
+    function updateStaking() public updateReward(msg.sender) {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            msg.sender
+        );
+        require(
+            node.id != 0 && node.active,
+            "No active node found for the staker address."
+        );
+
+        uint256 tokenId = users[msg.sender].tokenId;
+        require(tokenId != 0, "No staking found for the staker address.");
+
+        uint256 amount = valueOfBondedToken(tokenId);
+        require(
+            amount >= minStakeAmountPerNode,
+            "Insufficient amount to run a node."
+        );
+
+        uint64 tier = nodeManager.getTier(node.id);
+        uint256 maxStakeAmount = tiersMaxStakeAmount[tier];
+        if (amount > maxStakeAmount) {
+            amount = maxStakeAmount;
+        }
+
+        if (users[msg.sender].balance != amount) {
+            totalStaked -= users[msg.sender].balance;
+            users[msg.sender].balance = amount;
+            totalStaked += amount;
+            emit Staked(msg.sender, amount);
+        }
+    }
+
+    /**
+     * @dev Allows the stakers to withdraw their rewards.
+     * @param amount The amount of tokens to withdraw.
+     * @param reqId The id of the withdrawal request.
+     * @param signature A tss signature that proves the authenticity of the withdrawal request.
+     */
+    function getReward(
+        uint256 amount,
+        uint256 paidRewardPerToken,
+        bytes calldata reqId,
+        SchnorrSign calldata signature
+    ) public {
+        require(
+            !withdrawRequests[reqId],
```

Comment:
Is checking the duplicate reqIds necessary?
Users can simply send multiple requests to Muon and get multiple signed requests.

The contract should handle this on its own side.